lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1497,6 +1563,43 @@ Value FIRRTLLowering::getLoweredAndExtendedValue(Value value, Type destType) {
     return getOrCreateIntConstant(destWidth, 0);
   }
 
+  if (auto array = result.getType().dyn_cast<hw::ArrayType>()) {
+    // srcWidth == unsigned(destWidth) --> ok
+    // srcWidth > unsigned(destWidth)  --> error
+    // srcWidth < unsigned(destWidth)  --> insert concat/sext to elements
+
+    auto baseDestType = destType.cast<FIRRTLType>().getVectorBaseType();
+    auto baseSourceType =
+        value.getType().cast<FIRRTLType>().getVectorBaseType();
+    auto baseResultType = getVectorBaseType(result.getType());
+    if (!baseSourceType || !baseDestType || !baseResultType)
+      return {};
+
+    auto srcWidth = baseResultType.cast<IntegerType>().getWidth();
+    auto destWidth = baseDestType.cast<IntType>().getWidthOrSentinel();
+    if (destWidth == -1)
+      return {};
+
+    if (srcWidth == unsigned(destWidth))
+      return result;
+
+    if (srcWidth > unsigned(destWidth)) {
+      builder.emitError("operand should not be a truncation");
+      return {};
+    }
```

Comment:
I think the vector code here would be simpler if you build the logic into the map function instead of trying to decide on extending or truncating outside the map.  That would make the code be something like:

```
if (types already match) { // can use a pointer comparison here
  return done
}

auto extend = [&](Value value) {
      if (is a truncation) {
        emitError(...);
        return {};
      }
      if (value.getType() ... isSigned())
        return builder.createOrFold<comb::SExtOp>(resultType, value);
      auto zero = getOrCreateIntConstant(destWidth - srcWidth, 0);
      return builder.createOrFold<comb::ConcatOp>(zero, value);
  }
```
return mapArray(result, extend);
```

In addition to simplifying the code, this eliminates the need to dig into the elements to decide what to do.  I think this will work better when you add structs down the line.

Reply:
Thanks. I merged this logic into `mapArray` and renamed `mapArray` into `getExtOrTruncArrayValue`.